### PR TITLE
Updates with messages

### PR DESCRIPTION
**Pull request by rustyrussell**
Kind of went the long way around: channeld needs to ask gossipd for update_channel, and that's our first request-response.  The best way to treat this is to make channeld do gossipd comms sync (like it does for master comms); cleanest way to do that is to remove ccan/io, which also fixes another potential bug.

### CODE REVIEW

File: gossipd/gossip.c [hunk 1]
```diff
@@ -599,6 +599,51 @@ static struct io_plan *peer_start_gossip(struct io_conn *conn, struct peer *peer
...
+		update = NULL;
+		goto reply;
+	}
+
+	for (i = 0; i < tal_count(us->out); i++) {
+		if (!short_channel_id_eq(&us->out[i]->short_channel_id,
+					 &schanid))
+			continue;
+
+		update = us->out[i]->channel_update;
```
Comment (cdecker) [inline gossipd/gossip.c]: Maybe move this out of the loop down to `reply:`?

File: gossipd/gossip.c [hunk 2]
```diff
@@ -641,7 +641,7 @@ static void handle_get_update(struct peer *peer, const u8 *msg)
 
 reply:
 	msg = towire_gossip_get_update_reply(msg, update);
-	msg_enqueue(&peer->peer_out, take(msg));
+	daemon_conn_send(&peer->owner_conn, take(msg));
```
Comment (cdecker) [inline gossipd/gossip.c]: Doesn't this assume that it really is the owner asking for the update? Shouldn't we pass in the `daemon_conn` that the request came from, so we can also handle requests from other daemons?

File: channeld/channel.c [hunk 3]
```diff
@@ -1977,20 +1977,118 @@ static void handle_preimage(struct peer *peer, const u8 *inmsg)
 	abort();
 }
 
+static u8 *foreign_channel_update(const tal_t *ctx,
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply channeld/channel.c]: Yes...

File: gossipd/gossip.c (hunk 2 above)
Comment (rustyrussell) [reply gossipd/gossip.c]: Hmm, good point.

File: gossipd/gossip.c (hunk 1 above)
Comment (rustyrussell) [reply gossipd/gossip.c]: Yes... I was meaning to get back to this when you merged it anyway.  Oh well..

Comment (cdecker) [reply gossipd/gossip.c]: Sorry about that, I thought we can also address these things later and I should not delay this any further :-)

**Comment by rustyrussell** [comment]
Christian Decker <notifications@github.com> writes:
> +static void get_shared_secret(const struct htlc *htlc,
>
> I started hoisting up the function declarations instead of moving the function bodies. Are there other advantages moving bodies, besides not having two places that declare the function signature?

Yes, code moves should be done as a separate patch.

Generally, predeclaring static functions is wrong unless they're
mutually recursive and it can't be avoided.

Cheers,
Rusty.

**Comment by cdecker** [comment]
I did not know that, I'll keep that in mind for the future :-)

**Pull request merged**